File: 携程8.23笔试/t3.py
# 其实r没有必要做离散化,浪费了性能
# 我们只需要压缩l的值域即可.
# 这个版本的离散化写得有问题，应该输入全部读完以后统一离散化，否则两个部分的输入值域不同，离散化结果也是错误的
import sys
import bisect

input = sys.stdin.readline

N = int(5e5 + 5)
M = int(1e9 + 5)

class TreeArr:
    def __init__(self, len):
        self.len = len
        self.arr = [0] * (len + 1)
    
    def update(self, pos, x):
        while (pos <= self.len):
            self.arr[pos] += x
            pos += (pos & (-pos))
    
    def query(self, pos):
        res = 0
        while pos:
            res += self.arr[pos]
            pos -= (pos & (-pos))
        return res


def main():
    n, m = tuple(map(int, input().strip().split()))
    origin_vals = []
    bacs: list[tuple] = []
    for _ in range(n):
        l, r = tuple(map(int, input().strip().split()))
        origin_vals.append(l)
        origin_vals.append(r)
        bacs.append((l, r))
    # 离散化
    vals = sorted(set(origin_vals))
    _max = len(vals)
    func = lambda x: bisect.bisect_left(vals, x) + 1
    for i in range(len(bacs)):
        bacs[i] = (func(bacs[i][0]), func(bacs[i][1]))
    bacs = sorted(bacs, key=lambda x: x[1])
    
    origin_vals.clear()

    # 记录查询
    queries = []
    for i in range(m):
        l, r = tuple(map(int, input().strip().split()))
        queries.append((i, l, r))
        origin_vals.append(l)
        origin_vals.append(r)
    
    # 离散化
    for i in range(len(queries)):
        queries[i] = (queries[i][0], func(queries[i][1]), func(queries[i][2]))

    queries = sorted(queries, key=lambda x: x[2])

    tree_arr =TreeArr(_max)
    idx = 0
    ans = [0] * m

    # 计算答案
    for i in range(m):
        qid, ql, qr = queries[i]
        while idx < len(bacs) and bacs[idx][1] <= qr:
            tree_arr.update(bacs[idx][0], 1)
            idx += 1
        ans[qid] = idx - tree_arr.query(ql - 1)
    
    print("\n".join(map(str, ans)))

if __name__ == "__main__":
    main()

File: 携程8.23笔试/t3.cpp
// POJ 2352 Stars
#include <algorithm>
#include <array>
#include <cstdio>
#include <iterator>
#include <random>
#include <vector>

constexpr int N = 1e5 + 5, M = 1e9 + 5;
using PII = std::pair<int, int>;
std::vector<int> origin_vals;
std::vector<PII> bacs;
std::vector<std::tuple<int, int, int>> queries;

int n, m;

inline int lowbit(int x) {
    return x & (-x);
}

struct TreeArr {
    int len;
    std::vector<int> arr;
    TreeArr(int l): len(l) {
        arr.resize(len + 1, 0);
    }

    void update(int pos, int x) {
        while (pos <= len) {
            arr[pos] += x;
            pos += lowbit(pos);
        }
    }

    int query(int pos) {
        int res = 0;
        while (pos) {
            res += arr[pos];
            pos -= lowbit(pos);
        }
        return res;
    }
};

inline int mapping(std::vector<int> &a, int x) {
    return (std::lower_bound(a.begin(), a.end(), x) - a.begin()) + 1;
}

int main() {
    std::freopen("input.txt", "r", stdin);
    scanf("%d%d", &n, &m);
    int l, r;

    bacs.resize(n);
    queries.resize(m);

    for (int i = 0; i < n; i++) {
        scanf("%d%d", &l, &r);
        origin_vals.push_back(l);
        bacs[i] = {l, r};
    }

    for (int i = 0; i < m; i++) {
        scanf("%d%d", &l ,&r);
        origin_vals.push_back(l);
        queries[i] = std::tuple{r, l, i};
    }
    // 离散化
    std::sort(origin_vals.begin(), origin_vals.end());
    origin_vals.erase(std::unique(origin_vals.begin(), origin_vals.end()), origin_vals.end());
    for (int i = 0; i < n; i++) {
        bacs[i].first = mapping(origin_vals, bacs[i].first);
    }
    for (int i = 0; i < m; i++) {
        std::get<1>(queries[i]) = mapping(origin_vals, std::get<1>(queries[i]));
    }
    auto func = [](PII a, PII b)->bool {return a.second < b.second;};

    std::sort(bacs.begin(), bacs.end(), func);
    std::sort(queries.begin(), queries.end());

    auto tree_arr = TreeArr(origin_vals.size());
    int idx = 0;
    std::vector<int> ans(m, 0);
    for (int i = 0; i < m; i++) {
        while (idx < n && bacs[idx].second <= std::get<0>(queries[i])) {
            tree_arr.update(bacs[idx].first, 1);
            idx++;
        }
        int t = std::get<2>(queries[i]);
        ans[t] = idx - tree_arr.query(std::get<1>(queries[i]) - 1);
    }
    for (int i = 0; i < m; i++) 
        printf("%d\n", ans[i]);
    return 0;
}